src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java

```diff
@@ -0,0 +1,382 @@
+package org.broadinstitute.hellbender.engine.dataflow.datasources;
+
+import com.google.api.services.storage.Storage;
+import com.google.cloud.dataflow.sdk.Pipeline;
+import com.google.cloud.dataflow.sdk.runners.DataflowPipeline;
+import com.google.cloud.dataflow.sdk.transforms.Create;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.KV;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.genomics.dataflow.readers.bam.BAMIO;
+import com.google.cloud.genomics.dataflow.utils.GCSOptions;
+import com.google.cloud.genomics.utils.GenomicsFactory;
+import com.google.common.base.Stopwatch;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.SAMRecordIterator;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.ValidationStringency;
+import htsjdk.samtools.util.Locatable;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.broadinstitute.hellbender.engine.ReadsDataSource;
+import org.broadinstitute.hellbender.engine.dataflow.DoFnWLog;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.dataflow.BucketUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+
+import java.io.File;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * see ShardedReadsDatasource.get
+ */
+public final class ShardedReadsDatasource {
+    private final static Logger logger = LogManager.getLogger(ShardedReadsDatasource.class);
+
+    /**
+     * Create a {@link PCollection< KV <String,Iterable<GATKRead>>>} containing all the reads starting in the given intervals and their mates,
+     * grouped by the shard they start in.
+     * The shards are strings as per SimpleInterval's ToString syntax.
+     * Reads that are malformed or unmapped are ignored.
+     *
+     * Note: for performance reasons, this will give you SAMRecords that have no header. You have to put it back in.
```

Comment:
This warning needs to be way more prominent -- the reads will be returned in a potentially unusable state for some operations until the header is restored (at least until we patch htsjdk as planned in https://github.com/broadinstitute/hellbender/issues/900 to make headerless operation an officially supported mode).